cmd/kubeadm/app/phases/upgrade/compute_test.go

```diff
@@ -783,3 +783,50 @@ func TestGetBranchFromVersion(t *testing.T) {
 		})
 	}
 }
+
+func TestGetSuggestedEtcdVersion(t *testing.T) {
+	tests := []struct {
+		name              string
+		externalEtcd      bool
+		kubernetesVersion string
+		expectedVersion   string
+	}{
+		{
+			name:              "external etcd: no version",
+			externalEtcd:      true,
+			kubernetesVersion: "1.1.0",
+			expectedVersion:   "",
+		},
+		{
+			name:              "local etcd: illegal kubernetes version",
+			externalEtcd:      false,
+			kubernetesVersion: "1.x.5",
+			expectedVersion:   "N/A",
+		},
+		{
+			name:              "local etcd: no supported version for 1.10.5, return the nearest version",
+			externalEtcd:      false,
+			kubernetesVersion: "1.10.5",
+			expectedVersion:   "3.2.24",
+		},
+		{
+			name:              "local etcd: has supported version for 1.16.0",
+			externalEtcd:      false,
+			kubernetesVersion: "1.16.0",
+			expectedVersion:   "3.3.17-0",
+		},
+		{
+			name:              "local etcd: no supported version for v1.99.0, return the nearest version",
+			externalEtcd:      false,
+			kubernetesVersion: "v1.99.0",
+			expectedVersion:   "3.5.5-0",
```

Comment:
i think this is pinning our kubernetes version -> etcd version map in constants.go. which means we have update this new test every time we update the map. this would be undesired.

is that correct? if so, we might want to make getSuggestedEtcdVersion accept a custom map as a param.

Reply:
> 

right, i will add a custom map as a param, thanks